### PR TITLE
Suffix reorg

### PR DESCRIPTION
**Pull request by b00ris**
closes https://github.com/ledgerwatch/turbo-geth/issues/97

### CODE REVIEW

File: common/dbutils/suffix_type.go [hunk 1]
```diff
@@ -1,70 +1,65 @@
...
-	return Suffix(b)
+func Decode(b []byte) (SuffixHistory, error) {
+	if len(b) == 0 {
+		return SuffixHistory{
+			Changes: make([]Change, 0),
+		}, nil
+	}
+	buf := bytes.NewBuffer(b)
+	h := SuffixHistory{}
+	err := rlp.Decode(buf, &h)
```
Comment (AlexeyAkhunov) [inline common/dbutils/suffix_type.go]: I would prefer not to use `rlp` package, if possible. It is very inefficient both in terms of allocations and cpu usage. That is why I used hand-written encoding before

File: tests/pruner_test.go [hunk 2]
```diff
@@ -176,6 +176,163 @@ func TestBasisAccountPruning(t *testing.T) {
...
+		_, err = blockchain.InsertChain(types.Blocks{blocks[i]})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	res, err := getStat(db)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (b00ris) [inline tests/pruner_test.go]: Same for reorg

File: tests/pruner_test.go [hunk 3]
```diff
@@ -176,6 +176,163 @@ func TestBasisAccountPruning(t *testing.T) {
...
+	expected := stateStats{
+		NotFoundAccountsInHistory:     33,
+		ErrAccountsInHistory:          0,
+		ErrDecodedAccountsInHistory:   0,
+		NumOfChangesInAccountsHistory: 0,
+		AccountSuffixRecordsByTimestamp: map[uint64]uint32{
```
Comment (b00ris) [inline tests/pruner_test.go]: add comments for test

File: ethdb/database.go [hunk 4]
```diff
@@ -86,26 +86,40 @@ func (db *BoltDatabase) Put(bucket, key []byte, value []byte) error {
 }
 
 // Put puts the given key / value to the queue
-func (db *BoltDatabase) PutS(hBucket, key, value []byte, timestamp uint64) error {
+func (db *BoltDatabase) PutS(hBucket, key, value []byte, timestamp uint64, noHistory bool) error {
```
[the remark this answers is not in the document]
Comment (AlexeyAkhunov) [reply ethdb/database.go]: I think we can defer this change to: https://github.com/ledgerwatch/turbo-geth/issues/119

File: ethdb/table.go [hunk 5]
```diff
@@ -34,8 +34,8 @@ func (dt *table) Put(bucket, key []byte, value []byte) error {
 	return dt.db.Put(bucket, append([]byte(dt.prefix), key...), value)
 }
 
-func (dt *table) PutS(hBucket, key, value []byte, timestamp uint64) error {
-	return dt.db.PutS(hBucket, append([]byte(dt.prefix), key...), value, timestamp)
+func (dt *table) PutS(hBucket, key, value []byte, timestamp uint64, noHistory bool) error {
```
[the remark this answers is not in the document]
Comment (b00ris) [reply ethdb/table.go]: decided to keep it

File: tests/pruner_test.go (hunk 2 above)
Comment (b00ris) [reply tests/pruner_test.go]: Separated PR

File: tests/pruner_test.go (hunk 3 above)
Comment (b00ris) [reply tests/pruner_test.go]: done

**Pull request merged**